### DISCUSSION TITLE
failed to load tpcc common functions to Mysql

### DISCUSSION THREAD

**Discussion by tommysway** [TPC-Council/HammerDB#355] [General]
why ?

![image](https://user-images.githubusercontent.com/63028841/167365577-37588d65-0e59-4fc5-bd63-f9a6b7e59259.png)

**Comment by sm-shaw** [comment]
This is a duplicate of #350 please follow the troubleshooting guidance there.

**Reply by tommysway** [reply]
The error changed to :

Error in Virtual User 4: wrong # args: should be "do_tpcc host port socket count_ware user password db mysql_storage_engine partition num_vu"
Error in Virtual User 5: wrong # args: should be "do_tpcc host port socket count_ware user password db mysql_storage_engine partition num_vu"
Error in Virtual User 1: wrong # args: should be "do_tpcc host port socket count_ware user password db mysql_storage_engine partition num_vu"
Error in Virtual User 2: wrong # args: should be "do_tpcc host port socket count_ware user password db mysql_storage_engine partition num_vu"
Error in Virtual User 3: wrong # args: should be "do_tpcc host port socket count_ware user password db mysql_storage_engine partition num_vu"
Error in Virtual User 4: wrong # args: should be "do_tpcc host port socket count_ware user password db mysql_storage_engine partition num_vu"
Error in Virtual User 5: wrong # args: should be "do_tpcc host port socket count_ware user password db mysql_storage_engine partition num_vu"

**Reply by sm-shaw** [reply]
You probably have special characters in a field such as the password e.g. having a # in the line would comment out the rest of the line. To prevent this you need to use curly braces, escape with backslashes or an approach such as  #349 or if you don't need special characters then use plain text - in most test environments this should not be an issue.

**Reply by tommysway** [reply]
I changed the password, it's OK.

But a new question:

![image](https://user-images.githubusercontent.com/63028841/167589796-a730097e-90d4-4658-9a6d-a0c0c2ade245.png)

The first VUser is not done, is the whole create process  over ?

**Reply by sm-shaw** [reply]
No it is not, The first virtual user will create the tables, load the item table only and then typically wait for the rest to complete, once they are complete it will create the stored procedures and gather the statistics and end with the message "TPCC SCHEMA COMPLETE" (or similar). In this case it looks like it is still loading the item table which is unusual as this is 100,000 rows. It will report the rows loaded during the load so either wait for a while to see if it reports the first step or drop the database and retry. If the load is slow in general then this could be an indication of poor I/O performance both to the data area disk and also noting that for loading this way each insert is logged so redo performance is important to. Also look in the error log for any additional alerts.

**Reply by tommysway** [reply]
New question:

![image](https://user-images.githubusercontent.com/63028841/167628270-f9a8750a-a36f-4bc9-99b5-003e41c7e4ed.png)

**Reply by sm-shaw** [reply]
SSL is not enabled for MySQL connections in v4.4. PR #331 enables it for MariaDB in the next release. However, if you wish to use SSL and it is enabled on your database you can modify the connections in the script editor as it is supported in the underlying library just not directly in the GUI options.

**Reply by tommysway** [reply]
I set the skip_ssl and  it's OK.

But why cannot use the Metrics ? 

![image](https://user-images.githubusercontent.com/63028841/167753669-de5be12f-f786-4d85-87f9-2c5273cb6bea.png)

**Reply by sm-shaw** [reply]
The documentation is here https://www.hammerdb.com/docs/ch07.html - start the agent on the database server then in the Metrics options enter the hostname and id (port) reported by the agent. The metrics will then connect and show the CPU utilization.

**Reply by tommysway** [reply]
This can be accomplished by installing HammerDB on the SUT as well as the server.

Do you mean to install HammerDB's agent on the database server as well? And start the agent to get the performance data? Are there any installation instructions?

**Reply by sm-shaw** [reply]
Its all in the documentation. HammerDB is so lightweight you can install HammerDB in full on the SUT (ie the system where the database is running) or if you wish just copy bin, lib and agent. You then run agent on Linux or the agent batch file on Windows. 
you then see a message such as:
```
Initializing HammerDB Metric Agent 4.4
HammerDB Metric Agent active @ id 23760 hostname osprey (Ctrl-C to Exit)
```
This gives the id and hostname to put into the metrics options on the client

![metrics](https://user-images.githubusercontent.com/38044085/168032430-ea49f0f7-1d44-43dc-9a4a-2540bd40380c.PNG)

If it connects correctly (on windows you need defender permissions it will as for when starting) then you will see the CPU utilization.

![cpu](https://user-images.githubusercontent.com/38044085/168033028-f6bf2ad6-718e-40af-be45-ab528e938376.PNG)

 Pay close to attention to user (green) and system utilization (red) if there is a lot of red this can indicate system bottlenecks such as handling network interrupts.

**Reply by sm-shaw** [reply]
BTW - a windows client can connect to  a Linux agent and vice versa

**Reply by tommysway** [reply]
Yes, the agent is OK.

[root@mysql1 agent]# ./agent
Initializing HammerDB Metric Agent 4.4
HammerDB Metric Agent active @ id 11504 hostname mysql1 (Ctrl-C to Exit)

**Reply by tommysway** [reply]
![image](https://user-images.githubusercontent.com/63028841/168077937-92956d4b-38a7-4d2e-82a1-590b8410e90d.png)

**Comment by tommysway** [comment]
Why cannot load the postgresql library ?




[root@pg1 HammerDB-4.4]# yum list installed | grep postgre
postgresql.x86_64                     9.2.24-7.el7_9                   @updates
postgresql-contrib.x86_64             9.2.24-7.el7_9                   @updates
postgresql-libs.x86_64                9.2.24-7.el7_9                   @updates
postgresql-server.x86_64              9.2.24-7.el7_9                   @updates
[root@pg1 HammerDB-4.4]#
[root@pg1 HammerDB-4.4]# rpm -ql postgresql.x86_64 | grep lib64
/usr/lib64/pgsql
[root@pg1 HammerDB-4.4]#
[root@pg1 HammerDB-4.4]# export LD_LIBRARY_PATH=/usr/lib64/pgsql:$LD_LIBRARY_PATH
[root@pg1 HammerDB-4.4]#
[root@pg1 HammerDB-4.4]# ./hammerdbcli
HammerDB CLI v4.4
Copyright (C) 2003-2022 Steve Shaw
Type "help" for a list of commands
The xml is well-formed, applying configuration
hammerdb>librarycheck
Checking database library for Oracle
Error: failed to load Oratcl - can't read "env(ORACLE_HOME)": no such variable
Ensure that Oracle client libraries are installed and the location in the LD_LIBRARY_PATH environment variable
Checking database library for MSSQLServer
Error: failed to load tdbc::odbc - couldn't load file "libiodbc.so": libiodbc.so: cannot open shared object file: No such file or directory
Ensure that MSSQLServer client libraries are installed and the location in the LD_LIBRARY_PATH environment variable
Checking database library for Db2
Error: failed to load db2tcl - couldn't load file "/opt/HammerDB-4.4/lib/db2tcl2.0.0/libdb2tcl.so.0.0.1": libdb2.so.1: cannot open shared object file: No such file or directory
Ensure that Db2 client libraries are installed and the location in the LD_LIBRARY_PATH environment variable
Checking database library for MySQL
Error: failed to load mysqltcl - couldn't load file "/opt/HammerDB-4.4/lib/mysqltcl-3.052/libmysqltcl3.052.so": libmysqlclient.so.21: cannot open shared object file: No such file or directory
Ensure that MySQL client libraries are installed and the location in the LD_LIBRARY_PATH environment variable
Checking database library for PostgreSQL
Error: failed to load Pgtcl - couldn't load file "/opt/HammerDB-4.4/lib/pgtcl2.1.1/libpgtcl2.1.1.so": /opt/HammerDB-4.4/lib/pgtcl2.1.1/libpgtcl2.1.1.so: undefined symbol: lo_truncate64
Ensure that PostgreSQL client libraries are installed and the location in the LD_LIBRARY_PATH environment variable
Checking database library for MariaDB
Error: failed to load mariatcl - couldn't load file "/opt/HammerDB-4.4/lib/mariatcl0.1/libmariatcl0.1.so": libmariadb.so.3: cannot open shared object file: No such file or directory
Ensure that MariaDB client libraries are installed and the location in the LD_LIBRARY_PATH environment variable

hammerdb>quit
Shutting down HammerDB CLI
[root@pg1 HammerDB-4.4]#
[root@pg1 HammerDB-4.4]# ls -l /opt/HammerDB-4.4/lib/pgtcl2.1.1/libpgtcl2.1.1.so
-rwxr-xr-x. 1 root root 73688 Oct 17  2019 /opt/HammerDB-4.4/lib/pgtcl2.1.1/libpgtcl2.1.1.so
[root@pg1 HammerDB-4.4]#
[root@pg1 HammerDB-4.4]# ldd /opt/HammerDB-4.4/lib/pgtcl2.1.1/libpgtcl2.1.1.so
        linux-vdso.so.1 =>  (0x00007fff98b5f000)
        libpq.so.5 => /lib64/libpq.so.5 (0x00007fc2b8ed4000)
        libc.so.6 => /lib64/libc.so.6 (0x00007fc2b8b06000)
        libssl.so.10 => /lib64/libssl.so.10 (0x00007fc2b8894000)
        libcrypto.so.10 => /lib64/libcrypto.so.10 (0x00007fc2b8431000)
        libkrb5.so.3 => /lib64/libkrb5.so.3 (0x00007fc2b8148000)
        libcom_err.so.2 => /lib64/libcom_err.so.2 (0x00007fc2b7f44000)
        libgssapi_krb5.so.2 => /lib64/libgssapi_krb5.so.2 (0x00007fc2b7cf7000)
        libldap_r-2.4.so.2 => /lib64/libldap_r-2.4.so.2 (0x00007fc2b7a98000)
        libpthread.so.0 => /lib64/libpthread.so.0 (0x00007fc2b787c000)
        /lib64/ld-linux-x86-64.so.2 (0x00007fc2b9313000)
        libk5crypto.so.3 => /lib64/libk5crypto.so.3 (0x00007fc2b7649000)
        libdl.so.2 => /lib64/libdl.so.2 (0x00007fc2b7445000)
        libz.so.1 => /lib64/libz.so.1 (0x00007fc2b722f000)
        libkrb5support.so.0 => /lib64/libkrb5support.so.0 (0x00007fc2b701f000)
        libkeyutils.so.1 => /lib64/libkeyutils.so.1 (0x00007fc2b6e1b000)
        libresolv.so.2 => /lib64/libresolv.so.2 (0x00007fc2b6c01000)
        liblber-2.4.so.2 => /lib64/liblber-2.4.so.2 (0x00007fc2b69f2000)
        libsasl2.so.3 => /lib64/libsasl2.so.3 (0x00007fc2b67d5000)
        libssl3.so => /lib64/libssl3.so (0x00007fc2b6572000)
        libsmime3.so => /lib64/libsmime3.so (0x00007fc2b634a000)
        libnss3.so => /lib64/libnss3.so (0x00007fc2b6011000)
        libnssutil3.so => /lib64/libnssutil3.so (0x00007fc2b5de1000)
        libplds4.so => /lib64/libplds4.so (0x00007fc2b5bdd000)
        libplc4.so => /lib64/libplc4.so (0x00007fc2b59d8000)
        libnspr4.so => /lib64/libnspr4.so (0x00007fc2b579a000)
        libselinux.so.1 => /lib64/libselinux.so.1 (0x00007fc2b5573000)
        libcrypt.so.1 => /lib64/libcrypt.so.1 (0x00007fc2b533c000)
        librt.so.1 => /lib64/librt.so.1 (0x00007fc2b5134000)
        libpcre.so.1 => /lib64/libpcre.so.1 (0x00007fc2b4ed2000)
        libfreebl3.so => /lib64/libfreebl3.so (0x00007fc2b4ccf000)
[root@pg1 HammerDB-4.4]#

**Reply by sm-shaw** [reply]
The answer is here https://stackoverflow.com/questions/71252018/checking-postgresqul-library-from-hammerdb
install a PostgreSQL version from the test matrix in the documentation https://www.hammerdb.com/docs/ch01s04.html
the documentation can be really useful in finding answers to questions other people have asked before.

**Reply by tommysway** [reply]
Yes, I changed the OS and db version,then it's OK.
Thank you.

![image](https://user-images.githubusercontent.com/63028841/168101653-aeb0e909-4f7d-49a0-87b9-c1b6c0bedc8d.png)

**Reply by tommysway** [reply]
I would like to ask whether the cases of different types of database stress testing are logically identical. 

Because I found a huge difference in their test results. PostgreSQL has the best results, significantly an order of magnitude higher than other types of databases. 

I wonder if PG itself is superior or if the test case is different?

**Reply by sm-shaw** [reply]
This is a very open-ended question, as it depends a lot on the test environment.  Yes, the TPROC-C test is a derived workload from TPC-C so by its nature the NOPM value is intended to be comparable. The SQL is doing the same workload (and have all been adapted from the examples in the TPC-C specification), but not in an identical way. 
However, in general the commercial databases Oracle, SQL Server and Db2 offer the highest performance and scalability, this should not come as a major surprise as they have been tuned and optimized for TPC-C over many years - however the skillset to configure these databases for optimal performance may be higher. 
With the open-source databases yes PostgreSQL is able to take advantage of more advanced features such as bulk operations so can deliver higher throughput however the bloat can increase over time and the auotvacuum impact performance. 
Another general observation is that an expert DBA in one database who understands relational database architecture and concepts will typically be able to configure a database for higher throughput than a novice in another database, therefore it is not typically valid to make claims about superiority based on a limited number of test cases.
